### DISCUSSION TITLE
Something seems not quite right with the parsing production prologue and epilogue code

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#122] [Q&A]
This generated method from the C# grammar illustrates:
```java
    // CSharp.ccc:32:1
    final public void CompilationUnitBody() {
        if (cancelled) throw new CancellationException();
        String prevProduction = currentlyParsedProduction;
        this.currentlyParsedProduction = "CompilationUnitBody";
        // Code for BNFProduction specified at CSharp.ccc:32:1
        // top-level expansion ExpansionChoice
        // Code for ExpansionChoice specified at CSharp.ccc:33:5
        if (scan$CSharp_ccc$33$5()) {
            // choice for null index 0
            // Code for NonTerminal specified at CSharp.ccc:33:5
            pushOntoCallStack("CompilationUnitBody", "CSharp.ccc", 33, 5);
            try {
                FileScopedNamespaceDeclaration();
            } finally {
                popCallStack();
            }
        } else if (scan$CSharp_ccc$35$5()) {
            // choice for null index 1
            // Code for ZeroOrMore specified at CSharp.ccc:35:5
            while (true) {
                if (!(scan$CSharp_ccc$36$7())) break;
                // Code for NonTerminal specified at CSharp.ccc:38:7
                pushOntoCallStack("CompilationUnitBody", "CSharp.ccc", 38, 7);
                try {
                    Statement();
                } finally {
                    popCallStack();
                }
            }
            // Code for ZeroOrMore specified at CSharp.ccc:40:5
            while (true) {
                // Code for ExpansionChoice specified at CSharp.ccc:40:6
                if (scan$CSharp_ccc$40$6()) {
                    // choice for null index 0
                    // Code for NonTerminal specified at CSharp.ccc:40:6
                    pushOntoCallStack("CompilationUnitBody", "CSharp.ccc", 40, 6);
                    try {
                        NamespaceDeclaration();
                    } finally {
                        popCallStack();
                    }
                } else if (scan$CSharp_ccc$40$29()) {
                    // choice for null index 1
                    // Code for NonTerminal specified at CSharp.ccc:40:29
                    pushOntoCallStack("CompilationUnitBody", "CSharp.ccc", 40, 29);
                    try {
                        TypeDeclaration();
                    } finally {
                        popCallStack();
                    }
                } else {
                    break;
                }
            }
        } else {
            pushOntoCallStack("CompilationUnitBody", "CSharp.ccc", 33, 5);
            throw new ParseException(lastConsumedToken, CompilationUnitBody_FIRST_SET, parsingStack);
        }
    }
```
This contains the line `String prevProduction = currentlyParsedProduction;` in the prologue, but not the corresponding line `this.currentlyParsedProduction = prevProduction;` in the epilogue - which occurs in most of these methods. Now I see that the `CompilationUnitBody` has a different epilogue from the other productions, which invoke the `TreeBuildingAndRecovery` macro, which has the corresponding line. However, we shouldn't then be generating the line in the prologue.

**Comment by revusky** [comment]
It seems like the `popCallStack` method restores the `currentlyParsedProduction` variable to what it should be after the call, no? Though admittedly I am a bit confused about the code you point to. I'm wondering about the whole thing of saving/restoring `prevProduction`. Is that just redundant?

**Comment by vsajip** [comment]
I'm confused too - the `prevProduction` setup and restore has been there since I started working with JavaCC21, IIRC. Not sure if it's still needed! Also not sure about how `THIS_PRODUCTION` is used.

**Reply by revusky** [reply]
I think `THIS_PRODUCTION` was introuced by John Bradley.

**Reply by adMartem** [reply]
I realize this note is rather late, but just to close this loop, I first just changed the generated name for the top-level production node to `thisProduction` since I (and perhaps others) needed to reference the production in some cases where it is not the same as `CURRENT_NODE` (as when a node annotation is used and the `CURRENT_NODE` refers to that node).  I later added `THIS_PRODUCTION` as a moniker when I realized as I was syncing the non-Java languages that referencing it should be standardized regardless of language naming practices.  While it might seem that tree node annotations are normally not that common, in the case of synthetic nodes (necessary but not sufficient for JTB generation) they are ubiquitous and often generated many levels removed from the production they are contained within.

**Comment by vsajip** [comment]
OK - I've removed all references to `prevProduction` and all the tests still pass, so it'll be gone once I've merged my changes.

**Reply by revusky** [reply]
It looks like it is redundant. I think so, but look carefully and convince yourself of it. 

As a side note, I'd just say that I'm happy that you're in there learning the code in detail. It's not so terribly hard though it is more challenging to work with generated code because when there is a bug in generated code, you're always an extra degree away from the problem. When you look at the original JavaCC project, where they don't use templates, it is (a) quite impressive, in a way, that they could ever get it to work as well as it did, and (b) hardly surprising that they are terrified of going in there and doing any work on it!

**Comment by vsajip** [comment]
Well, not really learning it _effectively_, as for doing that, more of a roadmap would be needed. For example, I've just been looking at how the close condition for closing node scopes is computed, and it's not really in one place, so you have to hunt around and grep for stuff.
